probe-rs/src/architecture/arm/ap_v2/memory_access_port_interface.rs

```diff
@@ -0,0 +1,233 @@
+use crate::{
+    architecture::arm::{
+        ap_v2::registers::{DataSize, Register, CSW, DRW, TAR, TAR2},
+        communication_interface::SwdSequence,
+        memory::{ArmMemoryInterface, Status},
+        ApAddress, ArmError, ArmProbeInterface, DapAccess, FullyQualifiedApAddress,
+    },
+    probe::DebugProbeError,
+    MemoryInterface,
+};
+
+use super::registers::{BASE, BASE2};
+use super::MaybeOwned;
+
+pub struct MemoryAccessPortInterface<'iface> {
+    iface: MaybeOwned<'iface>,
+    base: u64,
+}
+impl<'iface> MemoryAccessPortInterface<'iface> {
+    /// creates a new `MemoryAccessPortInterface` from a reference to a `dyn ArmMemoryInterface`.
+    pub fn new_with_ref(
+        iface: &'iface mut (dyn ArmMemoryInterface + 'iface),
+        base: u64,
+    ) -> Result<Self, ArmError> {
+        // TODO! validity check from the parent root table
+        Ok(Self {
+            iface: MaybeOwned::Reference(iface),
+            base,
+        })
+    }
+
+    /// creates a new `MemoryAccessPortInterface` from a boxed `dyn ArmMemoryInterface`.
+    pub fn boxed(iface: Box<dyn ArmMemoryInterface + 'iface>, base: u64) -> Result<Self, ArmError> {
+        // TODO! validity check from the parent root table
+        Ok(Self {
+            iface: MaybeOwned::Boxed(iface),
+            base,
+        })
+    }
+
+    fn set_transaction_size(&mut self, size: DataSize) -> Result<(), ArmError> {
+        let mut csw_raw = [0u32];
+        self.iface
+            .read_32(self.base + u64::from(CSW::ADDRESS), &mut csw_raw)?;
+        let mut csw = CSW::try_from(csw_raw[0])?;
+        csw.SIZE = size;
+        csw.DbgSwEnable = true;
+
+        // TODO: This is setting the HPROT field to 0b10 and the HNONSEC = 1 for the AHB5 CSW. This
```

Comment:
These settings depend on the actual memory AP type. They're fine for now to get basic RP2350 support going, but in reality these should be abstracted away similar to how the ADIv5 implementation handles different buses.

Reply:
Makes sense to me.  I didn't notice the v5 handling when I wrote this code.

Reply:
I spawned https://github.com/probe-rs/probe-rs/issues/3061 to track the future unification here